prov/efa/test/efa_unit_test_cq.c

```diff
@@ -286,4 +287,122 @@ void test_rdm_cq_read_failed_poll()
+
+	/* Create and register a fake peer */
+	ret = fi_getname(&resource.ep->fid, &raw_addr, &raw_addr_len);
+	assert_int_equal(ret, 0);
+	raw_addr.qpn = 0; // return value of _read_qp_num
+	raw_addr.qkey = 0x1234;
+
+	if (!peer_removed) {
+		fi_addr_t peer_addr;
+		ret = fi_av_insert(resource.av, &raw_addr, 1, &peer_addr, 0, NULL);
+		assert_int_equal(ret, 1);
+	}
+
+	/* Setup packet entry */
+	entry = rxr_pkt_entry_alloc(rxr_ep, rxr_ep->efa_rx_pkt_pool, RXR_PKT_FROM_EFA_RX_POOL);
+	base_hdr.hdr.type = RXR_EAGER_MSGRTM_PKT; // packet type must be in [RXR_REQ_PKT_BEGIN, RXR_EXTRA_REQ_PKT_END)
+	base_hdr.hdr.flags |= RXR_PKT_CONNID_HDR | RXR_REQ_MSG;
+	base_hdr.hdr.msg_id = 0;
+	memcpy(entry->pkt, &base_hdr, sizeof(struct rxr_eager_msgrtm_hdr));
+	assert_int_equal(rxr_get_base_hdr(entry->pkt)->type, RXR_EAGER_MSGRTM_PKT);
+	assert_int_equal(rxr_pkt_req_base_hdr_size(entry), sizeof(struct rxr_eager_msgrtm_hdr));
+	opt_connid_hdr.connid = raw_addr.qkey;
+	memcpy(entry->pkt + sizeof(struct rxr_eager_msgrtm_hdr), &opt_connid_hdr, sizeof(struct rxr_req_opt_connid_hdr));
+	connid = rxr_pkt_connid_ptr(entry);
+	assert_int_equal(*connid, raw_addr.qkey);
+
+	if (!peer_removed) {
+		/* Skip handshake */
+		peer = rxr_ep_get_peer(rxr_ep, entry->addr);
+		assert_non_null(peer);
+		peer->flags |= RXR_PEER_HANDSHAKE_SENT;
+	}
+
+	/* Setup CQ */
+	efa_cq->ibv_cq_ex->wr_id = (uintptr_t)entry;
+	efa_cq->ibv_cq_ex->start_poll = &efa_mock_ibv_start_poll_return_mock;
+	efa_cq->ibv_cq_ex->next_poll = &efa_mock_ibv_next_poll_return_mock;
+	efa_cq->ibv_cq_ex->end_poll = &efa_mock_ibv_end_poll_return_mock;
+	efa_cq->ibv_cq_ex->read_slid = &efa_mock_ibv_read_slid_return_mock;
+	efa_cq->ibv_cq_ex->read_byte_len = &efa_mock_ibv_read_byte_len_return_mock;
+	efa_cq->ibv_cq_ex->read_opcode = &efa_mock_ibv_read_opcode_return_mock;
+	efa_cq->ibv_cq_ex->read_src_qp = &efa_mock_ibv_read_src_qp_return_mock;
 
+	efadv_cq.wc_read_ah = &efa_mock_efadv_wc_read_unknown_ah;
+
+	/* Read 1 entry with unknown AH */
+	will_return(efa_mock_ibv_start_poll_return_mock, 0);
+	will_return(efa_mock_ibv_next_poll_return_mock, ENOENT);
+	will_return(efa_mock_ibv_end_poll_return_mock, NULL);
+	will_return(efa_mock_ibv_read_slid_return_mock, 0xffff);
+	will_return(efa_mock_ibv_read_byte_len_return_mock, sizeof(base_hdr) + sizeof(opt_connid_hdr));
+	will_return_maybe(efa_mock_ibv_read_opcode_return_mock, IBV_WC_RECV);
+	will_return_maybe(efa_mock_ibv_read_src_qp_return_mock, raw_addr.qpn);
+
+	/*
+	 * Make sure to call efadv efadv_wc_read_ah, but
+	 * efadv_wc_read_ah is static inline and cannot be mocked directly
+	 */
+	g_efa_unit_test_mocks.efadv_cq_from_ibv_cq_ex = efadv_cq_from_ibv_cq_ex_return_mock;
+	expect_value(efadv_cq_from_ibv_cq_ex_return_mock, ibvcqx, efa_cq->ibv_cq_ex);
+	will_return(efadv_cq_from_ibv_cq_ex_return_mock, &efadv_cq);
+	will_return(efa_mock_efadv_wc_read_unknown_ah, raw_addr.raw);
+
+	ret = fi_cq_read(resource.cq, &cq_entry, 1);
+	assert_int_equal(ret, -FI_EAGAIN);
```

Comment:
This is not the case when peer is not removed. right? in this case we should get a completion entry.

Reply:
I will take a closer look.

Reply:
I think the reason is that `fi_recv` was not called, so in the "recover_forgotten" case, you will get an unexpected message. so we should check the unexpected message list here.